### PR TITLE
Hooks: on_exit, on_error

### LINKED ISSUE #81: Popup for rsync over ssh with password protected key
Currently there is no popup and probably only a failed exit code, would be great to forward a popup to the user.
I am not sure whether this is possible am not deep in the command processing yet.
If there was a callback function which is called when a key is not present in the running ssh agent, the user could open his own popup?

### PR DESCRIPTION
**Pull request by Dimfred**
Added two hooks, on_exit and on_error, to be called when rsync exits, or errors. This fixes #81

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -66,9 +66,15 @@ Global configuration settings with the default values
...
-    fugitive_sync = false
+    fugitive_sync = false,
     -- triggers `RsyncUp` when you save a file.
-    sync_on_save = true
+    sync_on_save = true,
+    -- called when the rsync command exits, provides the exit code and the used command
+    on_exit = function(code, command)
+    end,
+    -- called when the rsync command fails, provides the error message, and the used command
+    on_error = function(error_msg, command)
```
Comment (OscarCreator) [inline README.md:76]:
```suggestion
    -- called when the rsync command prints to stderr, provides the data and the used command
    on_stderr = function(data, command)
```

File: lua/rsync/config.lua [hunk 2]
```diff
@@ -2,6 +2,8 @@ local config = {}
 config.values = {
     fugitive_sync = false,
     sync_on_save = true,
+    on_exit = function(code, command)end,
+    on_error = function(error_msg, command)end,
```
Comment (OscarCreator) [inline lua/rsync/config.lua:6]:
```suggestion
    on_stderr = function(data, command) end,
```

File: lua/rsync/config.lua [hunk 3]
```diff
@@ -2,6 +2,8 @@ local config = {}
 config.values = {
     fugitive_sync = false,
     sync_on_save = true,
+    on_exit = function(code, command)end,
```
Comment (OscarCreator) [inline lua/rsync/config.lua:5]:
```suggestion
    on_exit = function(code, command) end,
```

File: lua/rsync/sync.lua [hunk 4]
```diff
@@ -30,6 +31,8 @@ local function safe_sync(command, on_start, on_exit)
             if vim.inspect(output) ~= vim.inspect({ "" }) then
                 log.info(string.format("safe_sync command: '%s', on_stderr: '%s'", command, vim.inspect(output)))
             end
+
+            config.values.on_error(output, command)
         end,
```
Comment (OscarCreator) [inline lua/rsync/sync.lua:37]:
```suggestion
        end,
        config.values.on_stderr(output, command)
```

**Review by OscarCreator — CHANGES_REQUESTED** [review]
Looks good, now just add test to verify this works as expected.

File: lua/rsync/sync.lua (hunk 4 above)
Comment (Dimfred) [reply lua/rsync/sync.lua:37]: Sry I think I don't understand you this then. Why would I put it out of the hook? This is a table so this won't work then?

Comment (OscarCreator) [reply lua/rsync/sync.lua:37]: Me neither, must have seen wrong. 😄

File: tests/rsync/rsync_spec.lua [hunk 5]
```diff
@@ -181,6 +182,60 @@ describe("rsync", function()
                 end)
             end)
         end)
+
+        describe("hooks", function()
+            it("success, on_exit called, on_error not called", function()
+                setup(function()
+                    Res_code = -1
+                    Res_data = ""
```
Comment (OscarCreator) [inline tests/rsync/rsync_spec.lua:190]:
```suggestion
                    res_data = ""
```

Use snake case

File: tests/rsync/rsync_spec.lua [hunk 6]
```diff
@@ -181,6 +182,60 @@ describe("rsync", function()
...
+                    assert.equals(Res_code, 0)
+                    assert.equals(Res_data[1], "")
+
+                    -- restore default
+                    require("rsync").setup({ on_exit = function(_, _) end, on_error = function(_, _) end })
+                end)
+            end)
+
+            it("error, on_exit called, on_error called", function()
+                setup(function()
```
Comment (OscarCreator) [inline tests/rsync/rsync_spec.lua:212]: Don't use setup here. It's not needed and you are overwriting the original file.

**Review by OscarCreator — CHANGES_REQUESTED** [review]

File: lua/rsync/config.lua [hunk 7]
```diff
@@ -3,6 +3,8 @@ config.values = {
     fugitive_sync = false,
     sync_on_save = true,
     project_config_path = ".nvim/rsync.toml",
+    on_exit = function(code, command) end,
+    on_error = function(data, command) end,
```
[the remark this answers is not in the document]
Comment (Dimfred) [reply lua/rsync/config.lua:7]: ahhh now I understand

File: tests/rsync/rsync_spec.lua (hunk 5 above)
Comment (Dimfred) [reply tests/rsync/rsync_spec.lua:190]: This is snake case, but the linter complains again, cause this is a global variable, global variables start with upper case.

Comment (OscarCreator) [reply tests/rsync/rsync_spec.lua:190]: Use `local res_data`

Comment (Dimfred) [reply tests/rsync/rsync_spec.lua]: Done

File: tests/rsync/rsync_spec.lua (hunk 6 above)
Comment (Dimfred) [reply tests/rsync/rsync_spec.lua]: done

**Review by OscarCreator — APPROVED** [review]

**Pull request merged**